---
title: "16S_SHP_test"
author: "Nat Pombubpa"
date: "Updated on November 1, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Documentation for Qiime2 Silva extracted length classifier on Sheephole crusts samples
This is a documentation for otu table, taxonomy table, and tree which were generated from Qiime2 using DADA2 and Silva database classifier extracted lenght (515-806). 

###STEP1: Load all necessary packages for analysis
More information about Phyloseq can be found at the following link: [Phyloseq](https://joey711.github.io/phyloseq/)
If you get error in this step, you probably need to install any packages which causes error.

```{r warning=FALSE, message=FALSE}
library(ape)
library(vegan)
library(plyr)
library(dplyr)
library(scales)
library(grid)
library(reshape2)
library(phyloseq)
library(magrittr)
library(ggplot2)
library(ggpubr)
library(data.table)
library(tidyr)
```

###STEP2: Transition data from Qiime2 to Phyloseq
Note: otu table and taxonomy table have been fixed right after Qiime2. Nat's Qiime2 script already include this step.
However, you can simply fix by changing "#Feature ID" to "OTU ID" or "OTU.ID" and the following commnand should work fine.

####STEP2.1: Import otu table from Qiime2

```{r}
otu = read.table(file="MojaveBacQ2FWLE.otu_table.fix.txt", header=T, sep='\t')
```

```{r}
head(otu)
```

Please note dimension of otu table. For this project, there are 21697 rows(OTU.ID) which will need to be matched to taxonomy table before starting Phyloseq analysis.
```{r}
dim(otu)
```

####STEP2.2: Import taxonomy table from Qiime2

```{r}
tax <- read.table(file="taxonomy.fix.tsv", sep='\t', header=TRUE)
head(tax)
```

Taxonomy table contains 22081 rows(OTU.ID) which is more than otu table. (Mitochondria and Chloroplast were removed in otu table which resulted in less number of OTU.ID in otu table.)
```{r}
dim(tax)
```

####STEP2.3: Match OTU.ID between otu and taxonomy table
Keep only OTU.ID in taxonomy table  that were found in otu table.
```{r}
tax_filtered <- tax[tax$OTU.ID %in% otu$OTU.ID,]
head(tax_filtered)
```

Check dimension of tax_filterd table: the number of OTU.ID should be the same as OTU table (in this case: 21697 OTU.ID)
```{r}
dim(tax_filtered)
```

Number of rows for "tax_filtered" and "otu" should be the same.
```{r}
dim(otu)
```

####STEP2.4: Select only OTU.ID column and taxonomy for taxonomy table
Taxonomy table from Qiime2 also contains the third column which is "Confidence". We will remove this column before loading data to Phyloseq.
```{r}
tax_filtered = tax_filtered[,c(1,2)]
```

After selection, taxonomy column will be separated into 7 levels as "Kingdom","Phylum","Class","Order", "Family", "Genus","Species"
```{r}
tax_filtered = separate(tax_filtered, Taxon, c("Kingdom","Phylum","Class","Order", "Family", "Genus","Species"), sep= ";", remove=TRUE)
```

At the end of this step, taxonomy table will contain OTU.ID column + 7 levels taxonomy columns (total of 8 columns). 
```{r}
dim(tax_filtered)
```

```{r}
head(tax_filtered)
```

Save "filtered" taxonomy table for Phyloseq and future usage.
```{r}
write.csv(tax_filtered, file="taxonomy_phyloseq_fix.csv")
```

###STEP3: Import otu table for analysis using Phyloseq 

```{r warning=FALSE}
otus <- read.table("MojaveBacQ2FWLE.otu_table.fix.txt",
                   header=T,sep="\t",row.names = 1)
otumat <- as(as.matrix(otus), "matrix")
OTU = otu_table(otumat, taxa_are_rows = TRUE)
```

###STEP4: Import taxonomy table for analysis using Phyloseq
This taxonomy table was created from STEP2.
```{r warning=FALSE}
taxmat <- read.csv("taxonomy_phyloseq_fix.csv", 
                   sep=",",row.names=1)
row.names(taxmat) = taxmat$OTU.ID
taxmat <- as(as.matrix(taxmat),"matrix")
TAX = tax_table(taxmat)
```

###STEP5: Import tree file for analysis using Phyloseq

```{r warning=FALSE}
treefile = "treeWMC.nwk"
tree = read.tree(treefile)
```

###STEP6: Import mapping file for analysis using Phyloseq

1.Check mapping file before import to R, R will automatically change sample name that starts with number or contain “-” in sample name. If you get error in this step, you should check sample name first. 

2.First column of first row should not start with #, R will not read the first row that starts with #

3. You can choose which samples to include in analysis by indicating specific group in the column

```{r warning=FALSE}
meta = read.table("Q2FWMojaveBacMetadata.tsv",
                  header=TRUE,row.names=1,
                  sep="\t",stringsAsFactors=FALSE)
```

Check if your metadata file has been import successfully and correctly, the output will show a table of your metadata file (mapping file). *If you do not have header, you might start your first row with # (remove # and reload your mapping file).
```{r}
tail(meta)
```

```{r}
meta = meta[which(meta$Description %in% c("Nicole")),]
```

```{r}
head(meta)
```

Construct sample_data-class using imported metadata
```{r warning=FALSE}
sampleData <- sample_data(meta)
```

###STEP7: Construct Phyloseq object
To construct phyloseq object, otu table, taxonomy table, and sampleData are required. Phylogenetic tree can be included, but it is not necessary for constructing phyloseq object.
Construct Phyloseq object called "Physeq"

```{r warning=FALSE}
physeq = phyloseq(OTU,TAX,sampleData,tree)
```

###STEP8: Check phyloseq object
This should indicate that your physeq is a "phyloseq-class experiment-level object"

```{r}
physeq
```

###STEP9: Remove singletons
Remove any OTUs that present only one time.

```{r}
physeq.prune = prune_taxa(taxa_sums(physeq) > 0, physeq)
```

###STEP10: Plot read counts to check dataset
Check read counts: any samples that have very low reads should be removed.
[Ref](http://evomics.org/wp-content/uploads/2016/01/phyloseq-Lab-01-Answers.html)

```{r}
readcount = data.table(as(sample_data(physeq.prune), "data.frame"),
                 TotalReads = sample_sums(physeq.prune), 
                 keep.rownames = TRUE)
setnames(readcount, "rn", "SampleID")

#For plotting, use command below.
#SeqDepth = ggplot(readcount, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth")
```

TotalReads of all the samples can be in this table (select only SampleID and TotalReads columns).
In order to check samples with low number of reads, "order()" can be used to sort "TotalReads" column.
In this dataset, N55.Rhizo has very low number of reads, so will will filter this sample out using the next minimum number of reads.
```{r}
readcount = readcount[order(readcount$TotalReads), c("SampleID", "TotalReads")]
```

```{r}
head(readcount)
```

###STEP11: Rarefy OTUs to a minimum number of reads
Rarefy OTUs (remove any samples that has very low number of reads)
If this step work fine, you have successfully imported data to R and completely generate phyloseq object as "physeq.prune.rarefy".
```{r warning=FALSE}
set.seed(711)
physeq.prune.rarefy = rarefy_even_depth(physeq.prune, sample.size = 43246, replace = FALSE, trimOTUs = TRUE)
physeq.prune.rarefy
```


###STEP12: Alpha Diversity plot

```{r}
plot_richness(physeq.prune.rarefy, x="Crust_type", color=("Crust_type"), measures=c("Observed")) + 
  geom_boxplot() + theme_bw() + ggtitle("Bacterial Alpha Diversity by Crust_type (Sheephole)") + 
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 0)) +
  stat_compare_means(method = "anova", label.y = 1580)
```

###STEP13: Calculate ordination 

```{r}
physeq.prune.ord <- ordinate(physeq.prune.rarefy, "PCoA", "unifrac")
```

###STEP14: Beta Diversity Plot
```{r }
plot_ordination(physeq.prune.rarefy, physeq.prune.ord, type = "samples", color = "Crust_type", shape = "Crust_type")  + theme_bw() + ggtitle("Bacterial Beta Diversity (PCoA) by Crust_type (Sheephole)") + theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(geom = "polygon", alpha = 1/8, aes(fill = Crust_type))
```








